### PR TITLE
Fix TimedeltaIndex +/- offset array

### PR DESCRIPTION
**Pull request by jbrockmendel**
No issue specific to this, but analogus to #18849.  There is a checkbox for this in #18824.

- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by jbrockmendel** [comment]
After this we'll be able to get rid of _TimeOp entirely

File: pandas/tests/indexes/timedeltas/test_arithmetic.py [hunk 1]
```diff
@@ -28,23 +29,98 @@ def freq(request):
...
-        expected = pd.TimedeltaIndex(['1 days 01:00:00', '3 days 04:02:00'])
+        tdi = TimedeltaIndex(['1 days 00:00:00', '3 days 04:00:00'])
+        other = np.array([pd.offsets.Hour(n=1), pd.offsets.Minute(n=-2)])
 
-        res = tdi + offs
+        expected = TimedeltaIndex([tdi[n] + other[n] for n in range(len(tdi))],
+                                  name=tdi.name, freq='infer')
+
+        with tm.assert_produces_warning(PerformanceWarning):
+            res = tdi + other
```
Comment (jreback) [inline pandas/tests/indexes/timedeltas/test_arithmetic.py]: shouldn't you be using the box here?

File: pandas/tests/indexes/timedeltas/test_arithmetic.py [hunk 2]
```diff
@@ -28,23 +29,98 @@ def freq(request):
...
+        anchored = Series([pd.offsets.MonthEnd(), pd.offsets.Day(n=2)],
+                          name=names[1])
+        with pytest.raises(TypeError):
+            with tm.assert_produces_warning(PerformanceWarning):
+                anchored + tdi
+        with pytest.raises(TypeError):
+            with tm.assert_produces_warning(PerformanceWarning):
+                tdi + anchored
+        with pytest.raises(TypeError):
+            with tm.assert_produces_warning(PerformanceWarning):
```
Comment (jreback) [inline pandas/tests/indexes/timedeltas/test_arithmetic.py]: some duplication here

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/indexes/timedeltas/test_arithmetic.py [hunk 3]
```diff
@@ -28,23 +29,98 @@ def freq(request):
...
-                             pd.offsets.Week(weekday=2)])
+        anchored = box([pd.offsets.QuarterEnd(),
+                        pd.offsets.Week(weekday=2)])
+        with pytest.raises(TypeError):
+            with tm.assert_produces_warning(PerformanceWarning):
+                tdi + anchored
+        with pytest.raises(TypeError):
+            with tm.assert_produces_warning(PerformanceWarning):
+                anchored + tdi
+
```
Comment (jreback) [inline pandas/tests/indexes/timedeltas/test_arithmetic.py]: ideally could parameterize some of these for index, ndarray & Series as well (you can leave them here is ok). I believe you have some duplication in test_base

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/indexes/timedeltas/test_arithmetic.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/tests/indexes/timedeltas/test_arithmetic.py]: Good catch.

File: pandas/tests/indexes/timedeltas/test_arithmetic.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/tests/indexes/timedeltas/test_arithmetic.py]: Yah, part of it is tricky because parametrizing with `box` and `names` at the same time is cumbersome.  My thought at the moment is to get the bugs fixed and tested first and take on refactoring later (when there's not a constant flow of rebasing to be done).

File: pandas/tests/indexes/timedeltas/test_arithmetic.py [hunk 4]
```diff
@@ -25,26 +28,88 @@ def freq(request):
     return request.param
 
 
+def _raises_and_warns(op, left, right):
```
Comment (jreback) [inline pandas/tests/indexes/timedeltas/test_arithmetic.py]: name things properly if you must do this: assert_raises_type_error_and_performance_warning.

yes its pretty long but its readable. remember folks who have no history with this code will be reading it and have no idea unless its spelled our.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jbrockmendel** [comment]
Travis error is a timeout; should I re-push?

**Comment by jreback** [comment]
i restarted

File: pandas/tests/indexes/timedeltas/test_arithmetic.py [hunk 5]
```diff
@@ -28,23 +29,95 @@ def freq(request):
...
+        # GH#18849
+        tdi = TimedeltaIndex(['1 days 00:00:00', '3 days 04:00:00'],
+                             name=names[0])
+        other = Series([pd.offsets.Hour(n=1), pd.offsets.Minute(n=-2)],
+                       name=names[1])
+
+        expected_add = Series([tdi[n] + other[n] for n in range(len(tdi))],
+                              name=names[2])
+
+        with tm.assert_produces_warning(PerformanceWarning):
```
Comment (jreback) [inline pandas/tests/indexes/timedeltas/test_arithmetic.py]: I would put a comment on before these sub-section of tests, hard to follow your flow others, IOW enumerate the cases. only looking for a comment at lines 89, 100, 107

**Review by jreback — CHANGES_REQUESTED** [review]
test comment, otherwise lgtm. ping when green.

File: pandas/tests/indexes/timedeltas/test_arithmetic.py (hunk 5 above)
Comment (jbrockmendel) [reply pandas/tests/indexes/timedeltas/test_arithmetic.py]: Will do.

**Comment by jbrockmendel** [comment]
Http error on appveyor

**Comment by jbrockmendel** [comment]
Ping

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!
